scipy/special/_hypergeometric.pxd

```diff
@@ -25,3 +30,53 @@ cdef inline double hyperu(double a, double b, double x) nogil:
             return poch(1.0 - b + a, -a)
 
     return hypU_wrap(a, b, x)
+
+
+cdef inline double hyp1f1(double a, double b, double x) nogil:
+    if npy_isnan(a) or npy_isnan(b) or npy_isnan(x):
+        return NPY_NAN
+    if b <= 0 and b == floor(b):
+        return NPY_INFINITY
+    elif a == 0 or x == 0:
+        return 1
+    elif a == -1:
+        return 1 - x / b
+    elif a == b:
+        return exp(x)
+    elif a - b == 1:
+        return (1 + x / b) * exp(x)
+    elif a == 1 and b == 2:
+        return (exp(x) - 1) / x
+    elif a <= 0 and a == floor(a):
+        # For `a` a negative integer the series is finite.
+        return hyp1f1_series(a, b, x)
+
+    if b > 0 and (fabs(a) + 1) * fabs(x) < 0.9 * b:
+        # At for the kth term of the series we are multiplying by
+        #
+        # t_k = (a + b) * x / ((b + k) * (k + 1))
+        #
+        # We have that
+        #
+        # |t_k| < (|a| + 1) * |x| / |b|,
+        #
+        # which means that in this branch we get geometric
+        # convergence.
+        return hyp1f1_series(a, b, x)
+
+    return hyp1f1_wrap(a, b, x)
+
+
+cdef inline double hyp1f1_series(double a, double b, double x) nogil:
+    cdef int k
+    cdef double term = 1
+    cdef double result = 1
+    for k in range(500):
```

Comment:
Does the 500 "magic number" come from the Fortran code (I didn't see it after quick look, but can barely read Fortran)?

And / or is it just some acceptable iteration count from the literature?

Reply:
The overall idea of using the series for this parameter regime is that it is roughly a geometric series with worst case ratio `0.9`. The relative error for that geometric series after summing `n` terms is

```
0.1 * 0.9**n / (1 - 0.9**(n + 1))
```

and at `n = 320` that is equal to `2.2782586118290245e-16`; i.e. about machine epsilon.

So `500` is an very rough, in theory overly-generous estimate of how long the series should need to converge in the worst-case scenario.

Reply:
So short answer-not from the Fortran source and not really from the literature; more a back-of-the-envelope estimate of how many terms we will need.

Reply:
Should the number be larger, because also the negative integer `a` branch ends up here?

Reply:
Good point; I ended up writing a separate function for the "`a` negative integer" case that does the error tracking so that the fast-convergence case isn't slowed down any. Ultimately we'll need a better algorithm for very negative `a`.

I also went with the (I think) slightly-tigher error estimate based on `sum |x_k| / |sum x_k|` from equation 2.6 of Higham, "The Accuracy of Floating Point Summation".